ethstorage/archiver/spec.md

```diff
@@ -0,0 +1,118 @@
+		"kzg_proof": "0x...",
+		},
+		...
+	]
+}
+```
+For comparison, the `/eth/v1/beacon/blob_sidecars/{block_id}` beacon API returns: 
+```json
+{
+	"data": [
+		{
+		"index": "0",
+		"blob": "0x...",
+		"signed_block_header": {
+		"message": {
+		"slot": "4648270",
+		"proposer_index": "1378",
+		"parent_root": "0x...",
+		"state_root": "0x...",
+		"body_root": "0x..."
+		},
+		"signature": "0x..."
+		},
+		"kzg_commitment": "0x...",
+		"kzg_proof": "0x...",
+		"kzg_commitment_inclusion_proof": [...]
+		},
+	...
+	]
+}
+```
+4. Es-node archiver should save all blobs managed by the EthStorage contract instead of only binding to Optimism's BatchInbox contract.
+5. Es-node archiver may need to provide the hash of the blob transaction for data verification, but this is depend on the clarification of requirements for Portal network integration.
+
+# Solutions
+
+While an HTTP server will be provided by es-node archiver as an external API service to let the consumer read data from `/eth/v1/beacon/blob_sidecars/{block_id}`, the method of retrieving metadata from L1 by es-node will involve different design approaches.
+
+## Solution 1: pre-stored
+
+1. Download:
+In downloader, the metadata (`kzg_commitment`, `kzg_proof`) can be downloaded with blob content through the `/eth/v1/beacon/blob_sidecars/{block_id}` beacon API,  and saved with kv index to leveldb as following:
+```
+BeaconRootHash->{
+		"data": [
+			{
+			"kv_index": 0,
+			"kzg_commitment": "0x...",
+			"kzg_proof": "0x...",
+			},
+			...
+			]
+		}
+```
+where `BeaconRootHash` can be queried from `/eth/v1/beacon/blocks/{block_id}/root`
+
+2. Sync:
+If the blob retention window has expired and the beacon API is no longer available, the metadata should be synced from peers. In p2p, the `BlobPayload` object will include `BeaconRootHash`, `kzg_commitment`, and `kzg_proof` that can be synced and stored to the key-value table mentioned above.
+
+3. Retrieve:
+It is convenient to load `kzg_commitment` and `kzg_proof` from the key-value . 
+To load blob data, we can we need to compute `dataHash` from `kzg_commitment` and pass `kvIdx` and `dataHash`(as `commit`)  to es-node storage manager:
+ ```
+ func (s *StorageManager) TryRead(kvIdx  uint64, readLen  int, commit  common.Hash) ([]byte, bool, error)
+ ```
+
+ ###  Considerations for Solution 1
+- There will be a flag to control the switch of the archive API, but this flag only controls the external reading interface,  while the downloading and syncing of archive data will always be on.
+- Blobs in the latest block are not supported to be downloaded immediately. You must wait for 2 epochs and L1 to download after the block is finalized.
+
+## Solution 2: on the fly
+
+1. From slot to EL block number:
+Consider `block_id` as the beacon chain slot number or block root, we can obtain the execution layer block number by querying `/eth/v2/beacon/blocks/{block_id}` and accessing `data.message.body.execution_payload.block_number`.
```

Comment:
Need to confirm that the `/eth/v2/beacon/blocks/{block_id}` still has valid data after 18 days

Reply:
Tested on sepolia with a slot in which blobs are supposed to be expired: https://sepolia.beaconcha.in/slot/4610252 (contains 2 blobs)

`curl -X 'GET'   'http://88.99.30.186:3500/eth/v1/beacon/blob_sidecars/4610252'   -H 'accept: application/json' `
returns `{"data":[]}`, while `curl -X 'GET'   'http://88.99.30.186:3500/eth/v2/beacon/blocks/4610252'   -H 'accept: application/json' ` still returns `execution_payload` data including block number and kzg commitments.

Reply:
Yes, the execution payload expiration will be done by EIP-4444.  We may also need to plan this in this doc (or at least put it in consideration).